src/index/hnsw/faiss_hnsw.cc

```diff
@@ -0,0 +1,1002 @@
+
+                    // set up local results
+                    faiss::idx_t* const __restrict local_ids = ids.get() + k * idx;
+                    float* const __restrict local_distances = distances.get() + k * idx;
+
+                    // set up faiss search parameters
+                    knowhere::SearchParametersHNSWWrapper hnsw_search_params;
+                    if (hnsw_cfg.ef.has_value()) {
+                        hnsw_search_params.efSearch = hnsw_cfg.ef.value();
+                    }
+                    // do not collect HNSW stats
+                    hnsw_search_params.hnsw_stats = nullptr;
+                    // set up feder
+                    hnsw_search_params.feder = feder_result.get();
+                    // set up kAlpha
+                    hnsw_search_params.kAlpha = bitset.filter_ratio() * 0.7f;
+
+                    // set up a selector
+                    BitsetViewIDSelector bw_idselector(bitset);
+                    faiss::IDSelector* id_selector = (bitset.empty()) ? nullptr : &bw_idselector;
+
+                    hnsw_search_params.sel = id_selector;
+
+                    // use knowhere-based search by default
+                    const bool override_faiss_search = hnsw_cfg.override_faiss_search.value_or(true);
+
+                    // check if we have a refine available.
+                    faiss::IndexRefine* const index_refine = dynamic_cast<faiss::IndexRefine*>(index.get());
+
+                    if (index_refine != nullptr) {
+                        // yes, it is possible to refine results.
+
+                        // cast a base index to IndexHNSW-based index
+                        faiss::IndexHNSW* const index_hnsw = dynamic_cast<faiss::IndexHNSW*>(index_refine->base_index);
+
+                        if (index_hnsw == nullptr) {
+                            // this is unexpected
+                            throw std::runtime_error("Expecting faiss::IndexHNSW");
+                        }
+
+                        // pick a wrapper for hnsw which does not own indices
+                        knowhere::IndexHNSWWrapper wrapper_hnsw_search(index_hnsw);
+                        knowhere::IndexBruteForceWrapper wrapper_bf(index_hnsw);
+
+                        faiss::Index* base_wrapper = nullptr;
+                        if (!override_faiss_search) {
+                            // use the original index, no wrappers
+                            base_wrapper = index_hnsw;
+                        } else if (whether_bf_search) {
+                            // use brute-force wrapper
+                            base_wrapper = &wrapper_bf;
+                        } else {
+                            // use hnsw-search wrapper
+                            base_wrapper = &wrapper_hnsw_search;
+                        }
+
+                        // check if used wants a refined result
+                        if (hnsw_cfg.refine_k.has_value()) {
```

Comment:
Milvus allows user to not provide any params, meaning any param we need to define its by default performance.

For this one: we set the `refine_k` as empty by default, and it means that the default performance for a user is not refine.

Do you think giving an experience number (can be ratio to k) to this `refine_k` to the refine as default is more intuitive? Also, we can define `refine_k = 0` as skipping refine, or even give a `skip_refine` param to make it more clear.

Reply:
This is the draft version. Someone who uses the system needs to let me know about the best way to expose params, the default way that Milvus users prefer.